serving/src/main/java/feast/auth/credentials/JwtAuthCredentials.java

```diff
@@ -0,0 +1,110 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package feast.auth.credentials;
+
+import static io.grpc.Metadata.ASCII_STRING_MARSHALLER;
+
+import io.grpc.CallCredentials2;
+import io.grpc.Metadata;
+import io.grpc.Status;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import javax.security.sasl.AuthenticationException;
+import okhttp3.FormBody;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import org.springframework.boot.configurationprocessor.json.JSONObject;
+
+/*
+ * Oauth Credentials Implementation for serving.
+ * Used by CoreSpecService to connect to core.
+ */
+public class JwtAuthCredentials extends CallCredentials2 {
+
+  static final String APPLICATION_JSON = "application/json";
+  static final String CONTENT_TYPE = "content-type";
+  static final String BEARER_TYPE = "Bearer";
+  static final String GRANT_TYPE = "grant_type";
+  static final String CLIENT_ID = "client_id";
+  static final String CLIENT_SECRET = "client_secret";
+  static final String AUDIENCE = "audience";
+  static final String OAUTH_URL = "oauth_url";
+  static final Metadata.Key<String> AUTHORIZATION_METADATA_KEY =
+      Metadata.Key.of("Authorization", ASCII_STRING_MARSHALLER);
+
+  private OkHttpClient httpClient;
+  private Request request;
+
+  public JwtAuthCredentials(CoreAuthenticationProperties authProperties) {
+    this.httpClient = new OkHttpClient();
+    Map<String, String> options = authProperties.getOptions();
+    if (!(options.containsKey(GRANT_TYPE)
+        && options.containsKey(CLIENT_ID)
+        && options.containsKey(AUDIENCE)
+        && options.containsKey(CLIENT_SECRET)
+        && options.containsKey(OAUTH_URL))) {
+      throw new AssertionError(
+          "please configure the properties:"
+              + " grant_type, client_id, client_secret, audience, oauth_url");
+    }
+    RequestBody requestBody =
+        new FormBody.Builder()
+            .add(GRANT_TYPE, options.get(GRANT_TYPE))
+            .add(CLIENT_ID, options.get(CLIENT_ID))
+            .add(CLIENT_SECRET, options.get(CLIENT_SECRET))
+            .add(AUDIENCE, options.get(AUDIENCE))
+            .build();
+    this.request =
+        new Request.Builder()
+            .url(options.get(OAUTH_URL))
+            .addHeader(CONTENT_TYPE, APPLICATION_JSON)
+            .post(requestBody)
+            .build();
+  }
+
+  @Override
+  public void thisUsesUnstableApi() {
```

Comment:
Should this be removed?

Reply:
no, it's unimplemented method in the abstract class. I believe it will be removed in future io.grpc releases.